balboa-client-jms/src/test/scala/com/blist/metrics/impl/queue/MetricJmsQueueSpec.scala

```diff
@@ -0,0 +1,177 @@
+      MetricRecord(1496268557542l, "3", "num-metadata-completed", 21,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "3", "num-metadata-available", 114,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "3", "num-metadata-total", 21,  Metric.RecordType.AGGREGATE),
+      MetricRecord(1496268557542l, "3", "num-metadata-total", 114,  Metric.RecordType.AGGREGATE),
+      MetricRecord(1496268557542l, "1", "num-metadata-completed", 5339,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "1", "num-metadata-available", 20916,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "1", "num-metadata-total", 5339,  Metric.RecordType.AGGREGATE),
+      MetricRecord(1496268557542l, "1", "num-metadata-total", 20916,  Metric.RecordType.AGGREGATE),
+      MetricRecord(1496268557542l, "14", "num-metadata-completed", 41,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "14", "num-metadata-available", 270,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "14", "num-metadata-total", 41,  Metric.RecordType.AGGREGATE),
+      MetricRecord(1496268557542l, "14", "num-metadata-total", 270,  Metric.RecordType.AGGREGATE)
+    )
+  }
+
+  abstract trait QueueSetup {
+    val mockQueue: Queue = mock[Queue]
+
+    val mockProducer: MessageProducer = mock[MessageProducer]
+
+    val mockSession: Session = mock[Session]
+
+    when(mockSession.createQueue("not-a-real-queue")).thenReturn(mockQueue)
+    when(mockSession.createProducer(mockQueue)).thenReturn(mockProducer)
+
+    val mockConnection: Connection = mock[Connection]
+
+    when(mockConnection.createSession(false, Session.AUTO_ACKNOWLEDGE)).thenReturn(mockSession)
+
+    case class MetricRecord(timestamp: Long, entityId: String, name: String, value: Number, recordType: Metric.RecordType)
+
+    def testMetrics: IndexedSeq[MetricRecord]
+
+    val metricWithZeroValue = new Metric()
+    metricWithZeroValue.setValue(0)
+    metricWithZeroValue.setType(Metric.RecordType.AGGREGATE)
+
+    def run() = {
+
+      lazy val bucketedMetrics = testMetrics.foldLeft(Map.empty[(Long, String), Metrics])((accum, metric) => {
+        accum.get((1496268480000l, metric.entityId)) match {
+          case Some(metrics: Metrics) => {
+            val metricObj = new Metric()
+
+            val value: Number = metric.recordType match {
+              case Metric.RecordType.ABSOLUTE => metric.value
+              case Metric.RecordType.AGGREGATE => {
+                val storedValue: Number = metrics.getOrDefault(metric.name, metricWithZeroValue).getValue
+                storedValue.intValue + metric.value.intValue
+              }
+              case _ => 0
+            }
+
+            metricObj.setValue(value)
+
+            metricObj.setType(metric.recordType)
+            accum ++ Map[(Long, String), Metrics]((metric.timestamp, metric.entityId) -> {
+              metrics.put(metric.name, metricObj)
```

Comment:
Blah.  We should work to make `Metrics` (and probably other objects used in here) immutable, probably just using scala collections under the hood.  For another day, though...